supercell-wx: init at 0.4.8

aware70 opened the pull request:
## Description of changes
Adds [Supercell Wx](https://supercell-wx.readthedocs.io/en/stable/#) at version 0.4.8.

Supercell Wx is a free, open source application to visualize live and archive NEXRAD Level 2 and Level 3 data, and severe weather alerts. It displays continuously updating weather data on top of a responsive map, providing the capability to monitor weather events using reflectivity, velocity, and other products. Extended functionality, including weather reports and lightning data can be added using placefiles.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

aware70 committed:
supercell-wx: init at 0.4.8

aware70 committed:
maintainers: add aware70

FliegendeWurst reviewed (changes requested):
Build fails:
```
FAILED: wxdata/CMakeFiles/wxdata.dir/source/scwx/provider/aws_level3_data_provider.cpp.o
/nix/store/vh9fsdhgxcnab2qk7vdp2palkkn6j3cp-gcc-wrapper-13.3.0/bin/g++ -DFMT_SHARED -DONLY_C_LOCALE=0 -DSPDLOG_COMPILED_LIB -DSPDLOG_FMT_EXTERNAL -DSPDLOG_SHARED_LIB -DUSE_OS_TZDB=1 -I/build/source/external/hsluv-c/src -I/build/source/wxdata/inclu
de -I/build/source/wxdata/source -I/build/source/external/units/include -isystem /nix/store/x6y7max0zx622sp0p12wi4xzmg82jw08-libxml2-2.13.4-dev/include/libxml2 -DBOOST_ALL_NO_LIB -O3 -DNDEBUG -std=c++20 -Wall -Wextra -Wpedantic -Werror -g -MD -MT
wxdata/CMakeFiles/wxdata.dir/source/scwx/provider/aws_level3_data_provider.cpp.o -MF wxdata/CMakeFiles/wxdata.dir/source/scwx/provider/aws_level3_data_provider.cpp.o.d -o wxdata/CMakeFiles/wxdata.dir/source/scwx/provider/aws_level3_data_provider.c
pp.o -c /build/source/wxdata/source/scwx/provider/aws_level3_data_provider.cpp
In file included from /nix/store/0vsyw5bhwmisszyfd1a0sdnwvnf4qa5a-gcc-13.3.0/include/c++/13.3.0/chrono:41,
                 from /build/source/wxdata/include/scwx/provider/nexrad_data_provider.hpp:5,
                 from /build/source/wxdata/include/scwx/provider/aws_nexrad_data_provider.hpp:3,
                 from /build/source/wxdata/include/scwx/provider/aws_level3_data_provider.hpp:3,
                 from /build/source/wxdata/source/scwx/provider/aws_level3_data_provider.cpp:3:
In constructor 'constexpr std::chrono::duration<_Rep, _Period>::duration(const _Rep2&) [with _Rep2 = long double; <template-parameter-2-2> = void; _Rep = long double; _Period = std::ratio<1>]',
    inlined from 'std::basic_istream<_CharT, _Traits>& date::from_stream(std::basic_istream<_CharT, _Traits>&, const CharT*, fields<Duration>&, std::__cxx11::basic_string<CharT, Traits, Alloc>*, std::chrono::minutes*) [with CharT = char; Traits =
std::char_traits<char>; Duration = std::chrono::duration<long int, std::ratio<1, 1000000000> >; Alloc = std::allocator<char>]' at /nix/store/4dm8l4hdhzm3bhiiixnp7wfa3wwps3c8-howard-hinnant-date-3.0.1-dev/include/date/date.h:7263:57,
    inlined from 'std::basic_istream<_CharT, _Traits>& date::from_stream(std::basic_istream<_CharT, _Traits>&, const CharT*, sys_time<Duration>&, std::__cxx11::basic_string<CharT, Traits, Alloc>*, std::chrono::minutes*) [with Duration = std::chron
o::duration<long int, std::ratio<1, 1000000000> >; CharT = char; Traits = std::char_traits<char>; Alloc = std::allocator<char>]' at /nix/store/4dm8l4hdhzm3bhiiixnp7wfa3wwps3c8-howard-hinnant-date-3.0.1-dev/include/date/date.h:8010:16,
    inlined from 'std::basic_istream<_CharT, _Traits>& date::operator>>(std::basic_istream<_CharT, _Traits>&, const parse_manip<Parsable, CharT, Traits, Alloc>&) [with Parsable = std::chrono::time_point<std::chrono::_V2::system_clock, std::chrono:
:duration<long int, std::ratio<1, 1000000000> > >; CharT = char; Traits = std::char_traits<char>; Alloc = std::allocator<char>]' at /nix/store/4dm8l4hdhzm3bhiiixnp7wfa3wwps3c8-howard-hinnant-date-3.0.1-dev/include/date/date.h:8080:23,
    inlined from 'static std::chrono::_V2::system_clock::time_point scwx::provider::AwsLevel3DataProvider::GetTimePointFromKey(const std::string&)' at /build/source/wxdata/source/scwx/provider/aws_level3_data_provider.cpp:121:35:
/nix/store/0vsyw5bhwmisszyfd1a0sdnwvnf4qa5a-gcc-13.3.0/include/c++/13.3.0/bits/chrono.h:574:17: error: 'S' may be used uninitialized [-Werror=maybe-uninitialized]
  574 |           : __r(static_cast<rep>(__rep)) { }
      |                 ^~~~~~~~~~~~~~~~~~~~~~~
In file included from /nix/store/4dm8l4hdhzm3bhiiixnp7wfa3wwps3c8-howard-hinnant-date-3.0.1-dev/include/date/tz.h:99,
                 from /build/source/wxdata/include/scwx/util/time.hpp:10,
                 from /build/source/wxdata/source/scwx/provider/aws_level3_data_provider.cpp:6:
/nix/store/4dm8l4hdhzm3bhiiixnp7wfa3wwps3c8-howard-hinnant-date-3.0.1-dev/include/date/date.h: In static member function 'static std::chrono::_V2::system_clock::time_point scwx::provider::AwsLevel3DataProvider::GetTimePointFromKey(const std::strin
g&)':
/nix/store/4dm8l4hdhzm3bhiiixnp7wfa3wwps3c8-howard-hinnant-date-3.0.1-dev/include/date/date.h:7261:37: note: 'S' declared here
 7261 |                         long double S;
      |                                     ^
cc1plus: all warnings being treated as errors
```
By the way, I think one commit `supercell-wx: init at 0.4.4` is sufficient for the package.

aware70 commented on pkgs/applications/science/meteorology/supercell-wx/default.nix:
```diff
@@ -0,0 +1,154 @@
+  meta = {
+    homepage = "https://supercell-wx.rtfd.io";
+    downloadPage = "https://github.com/dpaulat/supercell-wx/releases";
+    description = ''
```
Thanks for the input! I will work to address these.

aware70 replied:
@FliegendeWurst, I have:

- Rebased onto latest and squashed the package commits
- Moved the original description to `longDescription` and added a shorter `description`.
- Fixed the build errors. I'm not completely happy with part of the solution (add more `-Wno-error` flags) but I  think they will need to be addressed upstream. There is a new version available, `0.4.5`, that includes one of my patches but it looks like it will be another small project to update to it.

FliegendeWurst commented on pkgs/applications/science/meteorology/supercell-wx/default.nix:
```diff
@@ -0,0 +1,158 @@
+    ./patches/fix-cmake-install.patch
+    # }}}
+
+    # These may be submitted upstream {{{
```
Since once of these was accepted upstream, you can replace it with a fetchpatch of the relevant commit

aware70 replied:
Added fetchpatch for the audio codec fix.

FliegendeWurst commented on pkgs/applications/science/meteorology/supercell-wx/default.nix:
```diff
@@ -0,0 +1,158 @@
+  python3,
+  wrapQtAppsHook,
+}:
+let
```
I think it is preferred to have these in the derivation and use the `finalAttrs` pattern, to allow overriding.

aware70 replied:
I assume you mean `src`/`version` below these lines. I have made it so.

FliegendeWurst commented on pkgs/applications/science/meteorology/supercell-wx/patches/skip-git-versioning.patch:
```diff
@@ -0,0 +1,22 @@
+-        commitString  = commitString + "+dirty"
+-        copyrightYear = datetime.date.today().year
++    commitString = "@rev@"
++    copyrightYear = datetime.date.today().year
```
Not reproducible. I suppose it doesn't matter much though.

aware70 replied:
I could just hard-code it to the year of the revision being built.

aware70 replied:
I could also modify this script to read it from the LICENSE file. What do you think?

FliegendeWurst replied:
If it's not too much effort, then yes.

aware70 replied:
@FliegendeWurst I have updated this patch to read the copyright year from the license file.

FliegendeWurst commented on pkgs/applications/science/meteorology/supercell-wx/patches/use-find-package.patch:
```diff
@@ -0,0 +1,129 @@
+ project(scwx-data)
+ 
+-find_package(Boost)
++find_package(Boost REQUIRED COMPONENTS iostreams)
++find_package(BZip2)
+ find_package(cpr)
+ find_package(LibXml2)
+ find_package(re2)
+ find_package(spdlog)
++find_package(date)
++find_package(OpenSSL COMPONENTS Crypto)
```
So.. how does upstream build without these libraries? Just wondering

aware70 replied:
It doesn't - most of these are represented in the [conanfile](https://github.com/dpaulat/supercell-wx/blob/v0.4.4-release/conanfile.py). `BZip2` is acknowledged [here](https://github.com/dpaulat/supercell-wx/blob/v0.4.4-release/ACKNOWLEDGEMENTS.md?plain=1) but not explicitly tracked.

I bypassed the conanfile to bring in all the dependencies I could from Nix, and left alone some of the larger vendored dependencies (like `maplibre-native-qt`). Some of these might be here to fix linker errors for that.

The explicit link against `OpenSSL::Crypto` was to fix a linker error, but I could see that being a case of one of the Nix dependencies needing an `withOpenSSL = true` override somewhere.

FliegendeWurst commented on pkgs/applications/science/meteorology/supercell-wx/default.nix:
```diff
@@ -0,0 +1,158 @@
+    # These may be submitted upstream {{{
+    ./patches/add-cstdint.patch # use <cstdint> for GCC 13 compatibility
+    ./patches/fix-zoned-time.patch # fix ambiguity of some date/chrono functions in GCC 13
+    ./patches/fix-audio-codec-setup.patch # fix apparent logic error when checking for supported audio codecs on launch
```
It still claims the codecs are not installed. Is it working for you?

aware70 replied:
Ah, the patch included here missed a change to solve this issue more generally that was incorporated into the merged fix. Your other suggestion to fetch the patch directly may address this, although it could depend on your system configuration.

FliegendeWurst commented on pkgs/applications/science/meteorology/supercell-wx/default.nix:
```diff
@@ -0,0 +1,158 @@
+  curl,
+  glew,
+  geos,
+  git,
```
```suggestion
```

aware70 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -3854,6 +3854,8 @@ with pkgs;
 
   supergfxctl = callPackage ../applications/system/supergfxctl { };
 
+  supercell-wx = qt6.callPackage ../applications/science/meteorology/supercell-wx { };
```
Moved to `pkgs/by-name` and updated `qt6` references.

FliegendeWurst commented on pkgs/by-name/su/supercell-wx/package.nix:
```diff
@@ -0,0 +1,154 @@
+    maintainers = with lib.maintainers; [ aware70 ];
+  };
+
+  CXXFLAGS = "-Wno-error=restrict -Wno-error=maybe-uninitialized -Wno-error=deprecated-declarations";
```
```suggestion
  env.CXXFLAGS = "-Wno-error=restrict -Wno-error=maybe-uninitialized -Wno-error=deprecated-declarations";
```

FliegendeWurst commented on pkgs/by-name/su/supercell-wx/package.nix:
```diff
@@ -0,0 +1,154 @@
+    # fix apparent logic error when checking for supported audio codecs on launch
+    (fetchpatch {
+      name = "fix-audio-codec-setup.diff";
+      url = "https://patch-diff.githubusercontent.com/raw/dpaulat/supercell-wx/pull/223.diff";
```
Usually we do not reference a PR directly. The content of the patch may change if the PR is updated. Since this PR is merged, you can fetch the patches from the master branch.

FliegendeWurst commented on maintainers/maintainer-list.nix:
```diff
@@ -2221,6 +2221,11 @@
     githubId = 206242;
     name = "Andreas Wiese";
   };
+  aware70 = {
```
I think there is a merge conflict in this file

FliegendeWurst commented on pkgs/by-name/su/supercell-wx/package.nix:
```diff
@@ -0,0 +1,154 @@
+  curl,
+  glew,
+  geos,
+  boost185,
```
```suggestion
  boost,
```
boost 1.85 was removed some time ago. New default version is 1.86 so it should work. Also replace below.

aware70 replied:
I've made the suggested change.

FliegendeWurst commented:
Build fails on x64 Linux. Note that ofborg (and my build) are done by first merging the PR on current master. To reproduce the failure you need to rebase your branch on master.
```
FAILED: scwx-qt/CMakeFiles/scwx-qt.dir/source/scwx/qt/external/stb_image.cpp.o
/nix/store/xj72byimwa4w92f9vlrg921qdprr01vl-gcc-wrapper-14-20241116/bin/g++ -DBOOST_CONTAINER_DYN_LINK -DBOOST_CONTAINER_NO_LIB -DBOOST_IOSTREAMS_DYN_LINK -DBOOST_IOSTREAMS_NO_LIB -DBOOST_JSON_DYN_LINK -DBOOST_JSON_NO_LIB -DBOOST_TIMER_DYN_LINK ->
In file included from /build/source/scwx-qt/source/scwx/qt/external/stb_image.cpp:14:
/build/source/external/stb/stb_image.h: In function 'int stbi__parse_png_file(stbi__png*, int, int)':
/build/source/external/stb/stb_image.h:5166:56: error: writing 1 byte into a region of size 0 [-Werror=stringop-overflow=]
 5166 |                   for (k = 0; k < s->img_n; ++k) tc[k] = (stbi_uc)(stbi__get16be(s) & 255) * stbi__depth_scale_table[z->depth]; // non 8-bit images will be larger
      |                                                  ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/build/source/external/stb/stb_image.h:5082:25: note: at offset 3 into destination object 'tc' of size 3
 5082 |    stbi_uc has_trans=0, tc[3]={0};
      |                         ^~
cc1plus: all warnings being treated as errors
```

aware70 commented:
> Build fails on x64 Linux. Note that ofborg (and my build) are done by first merging the PR on current master. To reproduce the failure you need to rebase your branch on master.
> 
> ```
> FAILED: scwx-qt/CMakeFiles/scwx-qt.dir/source/scwx/qt/external/stb_image.cpp.o
> /nix/store/xj72byimwa4w92f9vlrg921qdprr01vl-gcc-wrapper-14-20241116/bin/g++ -DBOOST_CONTAINER_DYN_LINK -DBOOST_CONTAINER_NO_LIB -DBOOST_IOSTREAMS_DYN_LINK -DBOOST_IOSTREAMS_NO_LIB -DBOOST_JSON_DYN_LINK -DBOOST_JSON_NO_LIB -DBOOST_TIMER_DYN_LINK ->
> In file included from /build/source/scwx-qt/source/scwx/qt/external/stb_image.cpp:14:
> /build/source/external/stb/stb_image.h: In function 'int stbi__parse_png_file(stbi__png*, int, int)':
> /build/source/external/stb/stb_image.h:5166:56: error: writing 1 byte into a region of size 0 [-Werror=stringop-overflow=]
>  5166 |                   for (k = 0; k < s->img_n; ++k) tc[k] = (stbi_uc)(stbi__get16be(s) & 255) * stbi__depth_scale_table[z->depth]; // non 8-bit images will be larger
>       |                                                  ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> /build/source/external/stb/stb_image.h:5082:25: note: at offset 3 into destination object 'tc' of size 3
>  5082 |    stbi_uc has_trans=0, tc[3]={0};
>       |                         ^~
> cc1plus: all warnings being treated as errors
> ```

It looks like this was a new warning for gcc 14 from the updated stdenv. The same warning is present in the nixpkgs `stb`, which it turns out I wasn't actually using, so I fixed that too. I added a -Wno-error flag for it and fixed another gcc 14 include issue in the tests, which I will probably upstream.

However, it looks like a recent change to geographiclib (https://github.com/NixOS/nixpkgs/commit/90b3a6a25bbdd5f0db103a76a32d37da34f58813) which split the outputs has broken the project's integration with that library. Reverting the split outputs locally fixed it, but I will still investigate how to properly use the library in this new form.

aware70 commented:
Hey @sikmir, the project added by this PR uses `geographiclib` and appears to have a cmake issue since https://github.com/NixOS/nixpkgs/commit/90b3a6a25bbdd5f0db103a76a32d37da34f58813. I was able to work around it with
```nix
(geographiclib.overrideAttrs {
   outputs = [ "out" ];
})
```

Without that change, using
```cmake
find_package(GeographicLib)
```
in the project's cmake configuration produces this error:

```
> CMake Error at /nix/store/1vj70709rlmnldkgcxaj8nl78qa0iw5h-geographiclib-2.5-dev/lib/cmake/GeographicLib/geographiclib-targets.cmake:126 (message):
>   The imported target "GeographicLib::GeographicLib_SHARED" references the
>   file
>
>      "/nix/store/1vj70709rlmnldkgcxaj8nl78qa0iw5h-geographiclib-2.5-dev/lib/libGeographicLib.so.26.1.1"
>
>   but this file does not exist.  Possible reasons include:
>
>   * The file was deleted, renamed, or moved to another location.
>
>   * An install or uninstall procedure did not complete successfully.
>
>   * The installation package was faulty and contained
>
>      "/nix/store/1vj70709rlmnldkgcxaj8nl78qa0iw5h-geographiclib-2.5-dev/lib/cmake/GeographicLib/geographiclib-targets.cmake"
>
>   but not all the files it references.
>
> Call Stack (most recent call first):
>   /nix/store/1vj70709rlmnldkgcxaj8nl78qa0iw5h-geographiclib-2.5-dev/lib/cmake/GeographicLib/geographiclib-config.cmake:55 (include)
>   scwx-qt/scwx-qt.cmake:16 (find_package)
>   scwx-qt/CMakeLists.txt:8 (include)
>
>
> -- Configuring incomplete, errors occurred!
```

I think this may be related to: https://github.com/NixOS/nixpkgs/issues/144170.

sikmir commented:
> Hey @sikmir, the project added by this PR uses `geographiclib` and appears to have a cmake issue since [90b3a6a](https://github.com/NixOS/nixpkgs/commit/90b3a6a25bbdd5f0db103a76a32d37da34f58813). I was able to work around it with
> 
> ```nix
> (geographiclib.overrideAttrs {
>    outputs = [ "out" ];
> })
> ```
> 
> Without that change, using
> 
> ```cmake
> find_package(GeographicLib)
> ```
> 
> in the project's cmake configuration produces this error:
> 
> ```
> > CMake Error at /nix/store/1vj70709rlmnldkgcxaj8nl78qa0iw5h-geographiclib-2.5-dev/lib/cmake/GeographicLib/geographiclib-targets.cmake:126 (message):
> >   The imported target "GeographicLib::GeographicLib_SHARED" references the
> >   file
> >
> >      "/nix/store/1vj70709rlmnldkgcxaj8nl78qa0iw5h-geographiclib-2.5-dev/lib/libGeographicLib.so.26.1.1"
> >
> >   but this file does not exist.  Possible reasons include:
> >
> >   * The file was deleted, renamed, or moved to another location.
> >
> >   * An install or uninstall procedure did not complete successfully.
> >
> >   * The installation package was faulty and contained
> >
> >      "/nix/store/1vj70709rlmnldkgcxaj8nl78qa0iw5h-geographiclib-2.5-dev/lib/cmake/GeographicLib/geographiclib-targets.cmake"
> >
> >   but not all the files it references.
> >
> > Call Stack (most recent call first):
> >   /nix/store/1vj70709rlmnldkgcxaj8nl78qa0iw5h-geographiclib-2.5-dev/lib/cmake/GeographicLib/geographiclib-config.cmake:55 (include)
> >   scwx-qt/scwx-qt.cmake:16 (find_package)
> >   scwx-qt/CMakeLists.txt:8 (include)
> >
> >
> > -- Configuring incomplete, errors occurred!
> ```
> 
> I think this may be related to: #144170.

I will take a look.
